Util/Tokenization/src/main/java/org/tribuo/util/tokens/impl/wordpiece/WordpieceTokenizer.java

```diff
@@ -0,0 +1,205 @@
+        }
+        currentWordpieceIndex++;
+        if (currentWordpieceIndex < currentWordpieceTokens.size()) {
+            return true;
+        } else if (tokenizer.advance()) {
+            currentToken = this.tokenizer.getToken();
+            getWordpieceTokens();
+            currentWordpieceIndex = 0;
+            if (currentWordpieceTokens.size() == 0) {
+                return advance();
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private static String normalize(String text) {
+        text = Normalizer.normalize(text, Normalizer.Form.NFD);
+        text = text.replaceAll("\\p{Mn}", "");
+        return text;
+    }
+
+    private void getWordpieceTokens() {
+        this.currentWordpieceTokens.clear();
+
+        String text = currentToken.text;
+
+        if (neverSplitTokens.contains(text)) {
+            currentWordpieceTokens.add(currentToken);
+        }
+
+        if (toLowerCase) {
+            text = currentToken.text.toLowerCase();
+        }
+
+        if (this.stripAccents) {
+            text = normalize(text);
+        }
+
+        List<String> wordpieces = wordpiece.wordpiece(text);
+
+        if (wordpieces.size() == 0) {
+            return;
+        }
+
+        if (wordpieces.size() == 1) {
+            String wp = wordpieces.get(0);
+            if (wp.equals(this.wordpiece.getUnknownToken())) {
+                currentWordpieceTokens.add(new Token(wp, currentToken.start, currentToken.end, TokenType.UNKNOWN));
+            } else {
+                currentWordpieceTokens.add(new Token(wp, currentToken.start, currentToken.end, TokenType.WORD));
+            }
+            return;
+        }
+
+        int begin = currentToken.start;
+        for (String wp : wordpieces) {
+            TokenType type = TokenType.PREFIX;
+            int end = begin + wp.length();
+            if (wp.startsWith("##")) {
+                end -= 2;
+                type = TokenType.SUFFIX;
+            }
+            currentWordpieceTokens.add(new Token(wp, begin, end, type));
+            begin = end;
+        }
+    }
+
+    @Override
+    public Token getToken() {
+        if (currentWordpieceIndex < currentWordpieceTokens.size()) {
+            return currentWordpieceTokens.get(currentWordpieceIndex);
+        } else {
+            throw new IllegalStateException("WordpieceTokenizer is not ready.");
+        }
+    }
+
+    @Override
+    public String getText() {
+        return getToken().text;
+    }
+
+    @Override
+    public int getStart() {
+        return getToken().start;
+    }
+
+    @Override
+    public int getEnd() {
+        return getToken().end;
+    }
+
+    @Override
+    public TokenType getType() {
+        return getToken().type;
+    }
+
+    // TODO!
+    @Override
+    public WordpieceTokenizer clone() {
+        throw new RuntimeException(new CloneNotSupportedException());
```

Comment:
We'll need clone support for Tribuo, it clones the tokenizers in the text pipelines.

Reply:
implemented and tested.